src/fluree/db/json_ld/api.cljc

```diff
@@ -232,3 +233,39 @@
   [db query]
   (let [res-chan (query-api/query-async db query)]
     (promise-wrap res-chan)))
+
+(defn range
+  "Performs a range scan against the specified index using test functions
+  of >=, <=, >, <"
+  ;; TODO - assert index is valid index type
+  ([db index test match]
+   (promise-wrap
+     (query-range/index-range db index test match)))
+  ([db index start-test start-match end-test end-match]
+   (promise-wrap
+     (query-range/index-range db index start-test start-match end-test end-match))))
+
+(defn slice
+  "Like range, but returns all flakes that match the supplied flake parts."
+  [db index match]
+  (promise-wrap
+    (query-range/index-range db index = match)))
+
+(defn expand-iri
+  "Expands given IRI with the default database context, or provided context."
+  ([db compact-iri]
+   (db-proto/-expand-iri db compact-iri))
+  ([db compact-iri context]
+   (db-proto/-expand-iri db compact-iri context)))
+
+(defn internal-id
+  "Returns the internal Fluree integer id for a given IRI.
+  This can be used for doing range scans, slices and for other
+  more advanced needs.
+
+  Returns promise"
+  [db iri]
+  (promise-wrap
+    (->> (expand-iri db iri)
+         (db-proto/-subid db))))
+
```

Comment:
Do we really want these functions part of the public api? These are certainly useful functions but they expose our current indexing methods to the rest of the world and make a contract that we'll support it for the rest of time.

Reply:
Performing index range-scans is immensely more efficient and extremely common (we use it internally everywhere for that reason). Definitely an advanced user feature, and one we can caveat is dependent on the indexing strategy being used, but anyone being more advanced and needing performance-critical calls will gladly take this tradeoff.

Reply:
I think I would still like to keep this behind the `query` function - maybe we could add special implementation-specific query syntax for performing range scans manually. Putting this in the public api feels like giving up on a fast query planner - we should be able to figure out how to do efficient lookups on our own. And exposing this means we can't ever change our index structure without breaking someone else's program.

Reply:
Why conflate an existing API, arguably the most heavily used one (query), with a completely different way of retrieving data along with a different style output. This will make query harder for beginners while making the main use case of range scans - super fast lookups geared more for advanced users - less useful by at minimum being slower to do its task, and likely harder to use as I need to fumble through all the query documentation to find this unique syntax which most people already know by name: range scans.

As for the API changing, this is the most fundamental way of getting data. It has been a more stable API than the query API itself. It also aligns generally with range scan APIs that other databases feature, so I think anyone who is used to them will find the API what they generally expect.

Reply:
Sorry, didn't address query planner comment. An index range scan isn't a query, and thus there is nothing to plan, which is at least part of the reason they are so efficient. This is why some super fast query databases, e.g. Cassandra, are 100% based on index ranges - enabling them to completely avoid the complexity of query plans all together.

Reply:
Maybe instead we could put this in an impl namespace - still available to those in the know but distinct from the public api - you can use these functions but we don't promise not to break them in the future? That way if users want to use our internal "flake" abstraction they still can, but they'll be reminded that it's an internal implementation detail.

Reply:
I don't share the breaking API concern and I'm not sure where that concern originates. This is probably the least likely API to break - and as noted, the most commonly used API by Fluree itself... primarily for the reason that it is so useful. It also natively has polymorphism, in that as part of the API you specify the index you are scanning. If we have new index types, you'd simply specify those new index types - the API itself wouldn't need to change.

I also don't know why we'd want to hide a common database feature. I think its omission would be strange. 

We already allow user-specified query plans, in fact it is the only thing we allow.

Reply:
My understanding was that one of the main benefits and motivators of going to json-ld was removing our proprietary integer ids from the public api. That allows users to work with global IRIs, making data source open and combinable. A complementary idea with that was that anyone could build any sort of index using the transaction log. This change backtracks on that and I was confused by that.

Regarding the user-specified query plans, I meant the actual datastructure that we use when we do the analytical parse. I'm not sure what you mean by user-specified query plans - don't we take a query and then figure out the ranges ourselves? That in my mind is the query plan, though I'm not terribly familiar with the query pipeline or the specific terminology it uses.